akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala

```diff
@@ -689,6 +690,40 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
       Util.immutableSeq(f(elem))
     })
 
+  /**
+   * Transform each input element into a pair of an element of `T` and a state of `S`. The element will be emitted downstream
+   * and the state will be passed to the next step in the chain. The returned `T` MUST NOT be `null` as it is illegal
+   * as stream element - according to the Reactive Streams specification. For stateless variant see [[#map]].
+   *
+   * The initial state is created using the given function `create` when the flow is materializing.
+   *
+   * The `onComplete` function is called only once when the upstream or downstream finished, You can do some clean-up here,
+   * and if the returned value is not empty, it will be emitted to the downstream if available, otherwise the value will be dropped.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * '''Emits when''' the mapping function returns an element and downstream is ready to consume it
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam S the type of the state
+   * @tparam T the type of the output elements
+   * @param create a function that creates the initial state
+   * @param f a function that transforms the upstream element and the state into a pair of next state and output element
+   * @param onComplete a function that transforms the ongoing state into an optional output element
+   */
+  def statefulMap[S, T](create: function.Creator[S])(
+      f: function.Function[Pair[S, Out], Pair[S, T]],
+      onComplete: function.Function[S, Optional[T]]): javadsl.Flow[In, T, Mat] =
```

Comment:
That is accidental leaking of Scala API, it is anyways rendered as non-curried when used from Java but I think we can't fix it in case someone is using the curried form from Scala somewhere (we'd break source compatibility then).

Reply:
Thanks for clarify it, I will follow that style.